### PR TITLE
[IMP] base: support ordering on custom models

### PR DESCRIPTION
**Pull request by bouvyd**
Up until now, it was impossible to specify the default ordering
on models created manually.

This could somewhat be bypassed by specifying the ordering of records
on views themselves, but this has one main drawback: when using a
relational field that targets a custom model as a group-by key, the
ordering defaulted to the id of the custom record. For example, if I
create a custom field on partners that points to a custom model 'x_grade'
on which an 'x_sequence' field exists, I could order my grade in their
own list view according to their sequence, but any read on partners
grouped by this 'x_grade_id' field would order the returned groups by
id while I would prefer to have them ordered according to the
'x_sequence' field.

This commit introduces a new field 'default_order' on the ir.model model
that can store this default ordering clause (as an SQL expression).

### CODE REVIEW

**Comment by dbo-odoo** [comment]
Note: I named the field `model_order` because naming a column `order` is just asking for trouble

File: odoo/addons/base/models/ir_model.py [hunk 1]
```diff
@@ -99,6 +100,9 @@ def _default_field_id(self):
 
     name = fields.Char(string='Model Description', translate=True, required=True)
     model = fields.Char(default='x_', required=True, index=True)
+    default_order = fields.Char(string='Order', default='id',
```
Comment (rco-odoo) [inline odoo/addons/base/models/ir_model.py]: Why *default* order?  It looks misleading in this context.  I would simply name it `order`.

File: odoo/addons/base/models/ir_model.py [hunk 2]
```diff
@@ -156,6 +160,22 @@ def _check_model_name(self):
...
+        for model in self:
+            if self.env.get(model.model) is None:
+                # magic fields (like id) are not set up yet and registry
+                # is not updated, can't validate anything
+                # field is invisible in the view until first save + default is 'id'
+                # should be sufficient to assume things will go well
+                continue
+            stored_fields = list(map(lambda f: f.name, filter(lambda f: f.store == True, self.env[model.model]._fields.values())))
+            model._check_qorder(model.default_order)  # regex check for the whole clause ('is it valid sql?')
+            order_fields = re.findall(r'(?:\s*\"?([a-z0-9:_]+)\"?(?:\s+(?:desc|asc))?\s*(?:,|$))', model.default_order)
```
Comment (rco-odoo) [inline odoo/addons/base/models/ir_model.py]: The following regexp might be simpler: `r'"?(\w+)"?\s*(?:asc|desc)?'`and should be matched with `re.I`.

File: odoo/addons/base/views/ir_model_views.xml [hunk 3]
```diff
@@ -31,6 +31,7 @@
                             <field name="id" invisible="1"/>
                             <field name="name"/>
                             <field name="model" attrs="{'readonly': [('id', '!=', False)]}"/>
+                            <field name="default_order" attrs="{'required': [('id', '!=', False)], 'invisible': [('id', '=', False)]}"/>
```
Comment (rco-odoo) [inline odoo/addons/base/views/ir_model_views.xml]: Ha!  This is for the non-validated constraint, right?

**Review by rco-odoo — COMMENTED** [review]
Overall it looks good.

**Comment by dbo-odoo** [comment]
merged with another related PR #43981 for easier review

**Pull request closed without merging**